File: lingting-spring-redis/src/main/java/live/lingting/spring/redis/script/RedisScriptExecutor.java
package live.lingting.spring.redis.script;

import org.springframework.data.redis.connection.RedisConnection;
import org.springframework.data.redis.connection.RedisScriptingCommands;
import org.springframework.data.redis.connection.ReturnType;
import org.springframework.data.redis.core.RedisCallback;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.data.redis.core.script.DefaultScriptExecutor;
import org.springframework.data.redis.core.script.RedisScript;
import org.springframework.data.redis.serializer.RedisSerializer;

import java.util.List;

/**
 * @author lingting 2024-04-17 17:19
 */
@SuppressWarnings("unchecked")
public class RedisScriptExecutor<K> extends DefaultScriptExecutor<K> {

	protected final RedisTemplate<K, ?> template;

	public RedisScriptExecutor(RedisTemplate<K, ?> template) {
		super(template);
		this.template = template;
	}

	@Override
	public <T> T execute(RedisScript<T> script, RedisSerializer<?> argsSerializer, RedisSerializer<T> resultSerializer,
			List<K> keys, Object... args) {
		return template.execute((RedisCallback<T>) connection -> {
			ReturnType returnType = ReturnType.fromJavaType(script.getResultType());
			return eval(connection, returnType, argsSerializer, resultSerializer, scriptBytes(script), keys, args);
		});
	}

	public <T> T execute(RedisConnection connection, final RedisScript<T> script, final List<K> keys,
			final Object... args) {
		ReturnType returnType = ReturnType.fromJavaType(script.getResultType());
		return eval(connection, returnType, scriptBytes(script), keys, args);
	}

	public <T> T execute(RepeatRedisScript<T> script, List<K> keys, Object... args) {
		return template.execute((RedisCallback<T>) connection -> execute(connection, script, keys, args));
	}

	public <T> T execute(RedisConnection connection, RepeatRedisScript<T> script, List<K> keys, Object... args) {
		if (script.isLoad()) {
			return evalSha1(connection, script.getType(), script.getSha1(), keys, args);
		}
		RedisSerializer<String> serializer = template.getStringSerializer();
		byte[] bytes = serializer.serialize(script.getSource());
		return eval(connection, script.getType(), bytes, keys, args);
	}

	public <T> T eval(RedisConnection connection, ReturnType returnType, byte[] script, final List<K> keys,
			final Object... args) {
		return eval(connection, returnType, template.getValueSerializer(),
				(RedisSerializer<T>) template.getValueSerializer(), script, keys, args);
	}

	public <T> T eval(RedisConnection connection, ReturnType returnType, RedisSerializer<?> argsSerializer,
			RedisSerializer<T> resultSerializer, byte[] script, List<K> keys, Object... args) {
		byte[][] keysAndArgs = keysAndArgs(argsSerializer, keys, args);
		int keySize = keys.size();
		RedisScriptingCommands commands = connection.scriptingCommands();
		Object result = commands.eval(script, returnType, keySize, keysAndArgs);
		if (connection.isQueueing() || connection.isPipelined() || result == null) {
			return null;
		}
		return deserializeResult(resultSerializer, result);
	}

	public <T> T evalSha1(RedisConnection connection, ReturnType returnType, String sha1, final List<K> keys,
			final Object... args) {
		return evalSha1(connection, returnType, template.getValueSerializer(),
				(RedisSerializer<T>) template.getValueSerializer(), sha1, keys, args);
	}

	public <T> T evalSha1(RedisConnection connection, ReturnType returnType, RedisSerializer<?> argsSerializer,
			RedisSerializer<T> resultSerializer, String sha1, List<K> keys, Object... args) {
		byte[][] keysAndArgs = keysAndArgs(argsSerializer, keys, args);
		int keySize = keys.size();
		RedisScriptingCommands commands = connection.scriptingCommands();
		Object result = commands.evalSha(sha1, returnType, keySize, keysAndArgs);
		if (connection.isQueueing() || connection.isPipelined() || result == null) {
			return null;
		}
		return deserializeResult(resultSerializer, result);
	}

}

File: lingting-spring-redis/src/main/kotlin/live/lingting/spring/redis/script/RedisScriptExecutor.kt
package live.lingting.spring.redis.script

import com.fasterxml.jackson.databind.ObjectMapper
import live.lingting.framework.jackson.JacksonUtils
import live.lingting.framework.jackson.wrapper.JacksonWrapper
import live.lingting.framework.util.Slf4jUtils.logger
import live.lingting.spring.redis.Redis
import org.springframework.data.redis.connection.RedisScriptingCommands
import org.springframework.data.redis.core.RedisCallback
import org.springframework.data.redis.core.RedisTemplate
import org.springframework.data.redis.serializer.RedisSerializer

/**
 * @author lingting 2024-04-17 17:19
 */
class RedisScriptExecutor<T> @JvmOverloads constructor(
    val script: RepeatRedisScript<T>,
    val template: RedisTemplate<*, *> = Redis.instance().template(),
    val jacksonWrapper: JacksonWrapper<ObjectMapper> = JacksonUtils.jsonWrapper
) {

    companion object {
        val log = logger()
    }

    /**
     * 脚本内容序列化
     */
    var bytesSerializer = template.stringSerializer as RedisSerializer<Any>

    /**
     * 脚本key序列化
     */
    var keySerializer = template.keySerializer as RedisSerializer<Any>

    /**
     * 脚本参数序列化
     */
    var argsSerializer = template.valueSerializer as RedisSerializer<Any>

    /**
     * 脚本返回值序列化
     */
    var resultSerializer = template.valueSerializer as RedisSerializer<Any>

    fun load() {
        template.execute(RedisCallback {
            val commands = it.scriptingCommands()
            load(commands)
        })
    }

    fun load(commands: RedisScriptingCommands) {
        if (script.load) {
            return
        }
        val bytes = bytes()
        commands.scriptLoad(bytes)
        script.load = true
    }

    fun bytes(): ByteArray {
        if (script.bytes == null) {
            script.bytes = bytesSerializer.serialize(script.source)
        }
        return script.bytes!!
    }

    @JvmOverloads
    fun execute(keys: List<Any> = emptyList(), vararg args: Any): T? {
        return execute(keys, args.toList())
    }

    @JvmOverloads
    fun execute(keys: List<Any> = emptyList(), args: Collection<Any> = emptyList()): T? {
        if (script.load) {
            return evalSha1(keys, args)
        }
        return eval(keys, args)
    }

    @JvmOverloads
    fun eval(keys: List<Any> = emptyList(), vararg args: Any): T? {
        return eval(keys, args.toList())
    }

    @JvmOverloads
    fun eval(keys: List<Any> = emptyList(), args: Collection<Any> = emptyList()): T? {
        return commands(keys, args) { c, s, v ->
            val bytes = bytes()
            val eval = c.eval<T>(bytes, script.type, s, *v)
            script.load = true
            eval
        }
    }

    @JvmOverloads
    fun evalSha1(keys: List<Any> = emptyList(), vararg args: Any): T? {
        return evalSha1(keys, args.toList())
    }

    @JvmOverloads
    fun evalSha1(keys: List<Any> = emptyList(), args: Collection<Any> = emptyList()): T? {
        return commands(keys, args) { c, s, v ->
            c.evalSha(script.sha1, script.type, s, *v)
        }
    }

    fun <T> convert(value: Any, target: Class<T>?): T {
        try {
            if (target != null) {
                val convert = jacksonWrapper.convert(value, target)
                if (convert != null) {
                    return convert
                }
            }
        } catch (e: Exception) {
            log.debug("Convert error! value: {}; target: {}", value, target, e)
        }
        return value as T
    }

    fun keysAndArgs(keys: List<Any>, args: Collection<Any>): Array<ByteArray> {
        val list = mutableListOf<ByteArray>()

        val keyType = keySerializer.targetType
        for (key in keys) {
            val value = convert(key, keyType)
            val bytes = keySerializer.serialize(value)
            list.add(bytes!!)
        }

        val argType = argsSerializer.targetType
        for (arg in args) {
            val value = convert(arg, argType)
            val bytes = argsSerializer.serialize(value)
            list.add(bytes!!)
        }

        return list.toTypedArray()
    }

    fun deserializeResult(result: Any?): T? {
        if (result == null) {
            return null
        }

        if (result is ByteArray) {
            return resultSerializer.deserialize(result) as T?
        }

        if (result is Iterable<*>) {
            val list = mutableListOf<Any?>()
            for (item in result) {
                val deserialize = deserializeResult(item)
                list.add(deserialize)
            }
            return list as T?
        }

        return result as T?
    }

    fun commands(keys: List<Any>, args: Collection<Any>, callback: RedisScriptCallback): T? {
        val keysAndArgs = keysAndArgs(keys, args)

        return template.execute(RedisCallback {
            val commands = it.scriptingCommands()
            val apply = callback.apply(commands, keys.size, keysAndArgs)
            if (it.isQueueing || it.isPipelined || apply == null) {
                return@RedisCallback null
            }
            deserializeResult(apply)
        })
    }

    fun <T> copy(script: RepeatRedisScript<T>) = RedisScriptExecutor(script, template).let {
        it.bytesSerializer = bytesSerializer
        it.keySerializer = keySerializer
        it.argsSerializer = argsSerializer
        it.resultSerializer = resultSerializer
        it
    }

}
